File: Construct Binary Tree from Inorder and Postorder Traversal/Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.cpp
// 和上一道题目的解题思路是一样的
// 16ms 54.05%

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include<vector>
#include<algorithm>
using namespace std;

class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) 
    {
        return helper(0, postorder.size() - 1, 0, inorder.size() - 1, postorder, inorder);
    }

    TreeNode* helper(int postLeft, int postRight, int inLeft, int inRight, vector<int>& postorder, vector<int>& inorder)
    {   
        if (postLeft > postRight)
            return 0;
        else if (postLeft == postRight)
            return new TreeNode(postorder[postLeft]);
        else
        {
            int rootValue = postorder[postRight];
        
            vector<int>::iterator iter = find(inorder.begin() + inLeft, inorder.begin() + inRight + 1, rootValue);
            int rootIndex = iter - inorder.begin();

            TreeNode *root = new TreeNode(rootValue);
            root->left = helper(postLeft, postLeft + rootIndex - inLeft - 1, inLeft, rootIndex - 1, postorder, inorder);
            root->right = helper(postLeft + rootIndex - inLeft, postRight - 1, rootIndex + 1, inRight, postorder, inorder);

            return root;
        }
    }
};

File: Construct Binary Tree from Inorder and Postorder Traversal/Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.py
# 160ms 81.93%

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        return self.helper(0, len(postorder) - 1, 0, len(inorder) - 1, postorder, inorder)
        
    def helper(self, postLeft, postRight, inLeft, inRight, postorder, inorder):
        if postLeft > postRight:
            return None
        elif postLeft == postRight:
            return TreeNode(postorder[postRight])
        else:
            rootValue = postorder[postRight]
            rootIndex = inorder.index(rootValue)

            root = TreeNode(rootValue)
            root.left = self.helper(postLeft, postLeft + rootIndex - inLeft - 1, inLeft, rootIndex - 1, postorder, inorder)
            root.right = self.helper(postLeft + rootIndex - inLeft, postRight - 1, rootIndex + 1, inRight, postorder, inorder)

            return root
        